Initial version of slot-all element,

tomalec opened the pull request:
required for https://github.com/Starcounter/RebelsLounge/issues/368

tomalec commented:
@platosha, would you mind reviewing?

Do you know a way to workaround polyfill bug in tests?
`querySelector('foo > slot')` does not work in polyfilled browsers, as it seems compound selectors with slots does not match.

platosha commented:
@tomalec Indeed seems to be a bug in ShadyDOM.

The `children` list of the `#scope` element seems to be correct though, so I would suggest manual filtering of `children` as a workaround, for example:

```javascript
        it('should create `slot` elements for all slot names used by host\'s children, and `all-slotted` attribute to them', function() {
            const createdSlots = Array.from(host.shadowRoot.firstElementChild.children)
                .filter(el => el.matches('slot[name][all-slotted]'));
            expect(createdSlots).to.be.not.null;
            expect(createdSlots).to.have.length(2);
            Array.from(host.children).forEach((child)=>{
                child.hasAttribute('slot') &&
                    expect(createdSlots.find(el =>
                        el.name === child.getAttribute('slot')
                    )).to.be.not.null;
            });
        });
```

platosha commented on test/spec.html:
```diff
@@ -8,28 +8,146 @@
+        });
+
+        it('should create `slot` elements for all slot names used by host\'s children, and `all-slotted` attribute to them', function() {
+            debugger
```
Don’t forget to remove this `debugger`.

platosha commented on test/spec.html:
```diff
@@ -8,28 +8,146 @@
+            expect(host.shadowRoot.querySelectorAll(`#scope > slot[name][all-slotted]`)).to.have.length(2);
+            Array.from(host.children).forEach((child)=>{
+                child.hasAttribute('slot') &&
+                expect(host.shadowRoot.querySelector(`#scope > slot[name="${child.getAttribute('slot')}"][all-slotted]`)).to.be.not.null;
```
Hint: another option would be to check `child.assignedSlot`. Not a blocker though.

platosha commented on index.html:
```diff
@@ -1,21 +1,114 @@
     <script src="../webcomponentsjs/webcomponents-lite.js"></script>
 
+    <!-- Imports ace editor for nicer demo -->
+    <link rel="import" href="../juicy-ace-editor/juicy-ace-editor.html">
```
I’ve got 404 on this import. Should probably be added to `bower.json` as a dev dependency.

platosha commented:
What about the wildcard names `<slot-all name="myapp/*"></slot-all>`, btw? Is it coming later as a separate feature?

tomalec commented:
Thanks a lot for the review!
> What about the wildcard names `<slot-all name="myapp/*"></slot-all>`, btw? Is it coming later as a separate feature?

Yes, I wanted to start simplest possible, to unlock `uni-menu` work. Patterns will come later as a separate feature

tomalec commented:
It seems there is another ShadyDOM bug that heavily affects the functionality of the element, as mutation observer is not notified. @platosha have you faced such issue before?

i isolated the issue to https://github.com/webcomponents/shadydom/issues/260

platosha commented:
@tomalec Indeed can reproduce. I wasn’t aware of this issue, but looks like there is an `observeChildren` method in ShadyDOM just for that: https://github.com/webcomponents/shadydom/blob/7df92c0022eef22f2bc37d44da886cf250678585/tests/observeChildren.html#L116

Non-standard API :-(

tomalec commented:
Thanks! I was not aware of this `childrenObserver`. Unfortunately it solves half of the problem, as it does not trigger for children removal

tomalec commented:
I addressed the comments and managed to make it pass in polyfilled browser, but with a limitation, that it will not catch elements removed by `.remove`.

@warpech  do you have anything to add?

tomalec commented:
I addressed the comments and managed to make it pass in polyfilled browser, but with a limitation, that it will not catch elements removed by `.remove`.

@warpech  do you have anything to add?

warpech commented on slot-all.html:
```diff
@@ -0,0 +1,140 @@
+                        // if host still has the elements with the same slot name,
+                        // we should keep the slot
+                        const slot = __createdSlots.get(slotName);
+                        // Polyfileld browsers does not have assignedElements
```
`Polyfileld browsers does not`
↓
`Polyfilled browsers do not`

warpech commented on README.md:
```diff
@@ -29,44 +27,31 @@ Or [download as ZIP](https://github.com/Juicy/juicy-element/archive/master.zip).
-`--juicy-element-laser-color` | Color of the shooted lasers.
+WebComponentsJS Shadow DOM polyfill has some limitations.
+It does ignore `node.remove()` and `.insertAdjacentElement()` calls,
+meaning it cannot be observed or reflected in shadow root. This result in `slot-all` element being unable to react on child nodes been added or removed using this methods. If you are targeting polyfilled environments try using `node.parentNode.removeChild(node)` and `insertBefore`.
```
`this methods` -> `these methods`

warpech commented on index.html:
```diff
@@ -1,21 +1,114 @@
+    </div>
+
+
+    <h4>Looking for a more specs? Check out the <a href="test/">test suite</a>.</h4>
```
`Looking for a more specs?` -> `Looking for more specs?`

warpech commented on index.html:
```diff
@@ -1,21 +1,114 @@
+    // bound editor to the example
+    document.querySelector('juicy-ace-editor').addEventListener('change', (ev)=>{
+        document.querySelector('[shadow-host]').innerHTML = ev.target.value;
+        // trick Ace editor to adatpt to the new content
```
`adatpt` -> `adapt`

warpech commented on slot-all.html:
```diff
@@ -0,0 +1,140 @@
+            }
+            /**
+             * Remove slots for given elements, if needed:
+             *  - there is no more host's children with a slot name for given slot name
```
`there is no more host's children with a slot name for given slot name`

↓

`there are no more host's children with a slot name for the given slot name`

warpech commented on test/spec.html:
```diff
@@ -8,28 +8,161 @@
+            expect(host.shadowRoot.querySelector('slot-all')).to.be.an.instanceOf(customElements.get('slot-all'));
+        });
+
+        it('should create `slot` elements for all slot names used by host\'s children, and `all-slotted` attribute to them', function() {
```
> and `all-slotted`

↓

> and add `all-slotted`
